framework/include/utils/Enumerate.h

```diff
@@ -0,0 +1,138 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef ENUMERATE_H
+#define ENUMERATE_H
+
+#include <iterator>
+#include <iostream>
+
+/**
+ * Enumerate function for iterating over a range and obtaining both a reference to the underlying
+ * type and an index simultaneously. This method is forward-compatible with the C++17 structured
+ * bindings capability.
+ *
+ * C++11 compatible usage:
+ *
+ * for (auto it : enumerate(values))
+ *   _console << it.index() << ": " << it.value() << '\n';
+ *
+ * // Here the third argument is the starting index value
+ * for (auto it : enumerate(values.begin(), values.end(), 0))
+ *   _console << it.index() << ": " << it.value() << '\n';
```

Comment:
If you are trying to show how to use your macro, this might need the namespace, because that's what you want people to do.

Reply:
Done